lib/bolt/transport/docker/connection.rb

```diff
@@ -8,60 +8,85 @@ module Transport
+          raise "Could not find a container with name or ID matching '#{@target.host}'" if index.nil?
+          # Now find the indepth container information
+          output = execute_local_docker_json_command('inspect', [output[index]["ID"]])
+          # Store the container information for later
+          @container_info = output[0]
           @logger.debug { "Opened session" }
+          true
         rescue StandardError => e
           raise Bolt::Node::ConnectError.new(
             "Failed to connect to #{@target.uri}: #{e.message}",
             'CONNECT_ERROR'
           )
         end
 
+        # Executes a command inside the target container
+        #
+        # @param command [Array] The command to run, expressed as an array of strings
+        # @param options [Hash] command specific options
+        # @option opts [String] :interpreter statements that are prefixed to the command e.g `/bin/bash` or `cmd.exe /c`
+        # @option opts [Hash] :environment A hash of environment variables that will be injected into the command
+        # @option opts [IO] :stdin An IO object that will be used to redirect STDIN for the docker command
         def execute(*command, options)
           command.unshift(options[:interpreter]) if options[:interpreter]
+          # Build the `--env` parameters
+          envs = []
           if options[:environment]
-            envs = options[:environment].map { |env, val| "#{env}=#{val}" }
-            command = ['env'] + envs + command
+            options[:environment].each { |env, val| envs.concat(['--env', "#{env}=#{val}"]) }
           end
 
-          @logger.debug { "Executing: #{command}" }
-          result = @container.exec(command, options) { |stream, chunk| @logger.debug("#{stream}: #{chunk}") }
+          command_options = []
+          # Need to be interactive if redirecting STDIN
+          command_options << '--interactive' unless options[:stdin].nil?
+          command_options.concat(envs) unless envs.empty?
+          command_options << container_id
+          command_options.concat(command)
+
+          @logger.debug { "Executing: exec #{command_options}" }
+
+          result = execute_local_docker_command('exec', command_options, options[:stdin])
+
+          # The actual result is the exitstatus not the process object
+          result[2] = result[2].nil? ? -32768 : result[2].exitstatus
           if result[2] == 0
             @logger.debug { "Command returned successfully" }
           else
             @logger.info { "Command failed with exit code #{result[2]}" }
           end
-          result[0] = result[0].join.force_encoding('UTF-8')
-          result[1] = result[1].join.force_encoding('UTF-8')
+          result[0] = result[0].force_encoding('UTF-8')
+          result[1] = result[1].force_encoding('UTF-8')
+          # Normalise line endings
+          result[0].gsub!("\r\n", "\n")
+          result[1].gsub!("\r\n", "\n")
           result
```

Comment:
And then instead of `result`, rebuild the array like

```ruby
[stdout_str, stderr_str, status]
```

Reply:
Done.